src/main/java/edu/kit/datamanager/security/filter/KeycloakTokenFilter.java

```diff
@@ -0,0 +1,181 @@
+        boolean authenticated = false;
+        try {
+            Jws<Claims> claimsJws = tokenValidator.getJwsClaims(token);
+            Set<Map.Entry<String, Object>> claims = claimsJws.getBody().entrySet();
+            Map<String, Object> claimMap = new HashMap<>();
+            claims.forEach((entry) -> {
+                claimMap.put(entry.getKey(), entry.getValue());
+            });
+            jwToken = JwtAuthenticationToken.factoryToken(token, claimMap);
+
+            if (jwToken instanceof JwtServiceToken && ((JwtServiceToken) jwToken).getSources() != null) {
+                JwtServiceToken serviceToken = (JwtServiceToken) jwToken;
+                LOG.debug("Performing source check for JWToken for service {} and sources {}.", serviceToken.getPrincipal(), Arrays.asList(serviceToken.getSources()));
+                String remoteAddr = request.getRemoteAddr();
+                LOG.debug("Trying to match remote address {} with at least one allowed source.", remoteAddr);
+                boolean matchFound = false;
+                for (String source : serviceToken.getSources()) {
+                    if (NetworkUtils.matches(remoteAddr, source)) {
+                        matchFound = true;
+                        break;
+                    }
+                }
+                if (!matchFound) {
+                    LOG.warn("Invalid request from remote address {} to service {} found. Request denied.", remoteAddr, serviceToken.getPrincipal());
+                    throw new InvalidAuthenticationException("You are not allowed to authenticate using the provided token from your current location.");
+                }
+            }
+            LOG.trace("JWT validation finished. Checking result.");
+            if (jwToken != null && SecurityContextHolder.getContext().getAuthentication() == null) {
+                LOG.info("Authenticated username: {}", jwToken.getPrincipal());
+                setContext(request, jwToken);
+            } else {
+                try {
+                    LOG.info("Invalid Request: Token is expired or tampered");
+                    response.sendError(HttpServletResponse.SC_UNAUTHORIZED, "Unauthorized: Token is expired or tampered");
+                } catch (IOException ex) {
+                    LOG.error("Failed to send Unauthorized response back to client.", ex);
+                }
+            }
+            authenticated = true;
+        } catch (ExpiredJwtException ex) {
+            LOG.debug("Provided token has expired. Refresh of login required.", ex);
+            throw new InvalidAuthenticationException("Your token has expired. Please refresh your login.");
+        } catch (MalformedJwtException ex) {
+            LOG.debug("Provided token is malformed.", ex);
+            throw new MalformedJwtException("Your token is malformed.");
+        } catch (IllegalArgumentException ex) {
+            LOG.debug("Illegal argument exception while local authentication attempt.", ex);
```

Comment:
No, because this exception is thrown only in case of a misconfiguration. If the local key is shorter than the required length for the selected signature algorithm, local authentication will fail but this should not be returned to the client. Instead, authentication using keycloack should be tried directly afterwards as this is the primary authentication.